File: golang-devildyw/middle/L17电话号码的字母组合.go
package main

func main() {}

func letterCombinations(digits string) []string {
	var paths []string
	m := map[byte][]byte{
		'2': {'a', 'b', 'c'},
		'3': {'d', 'e', 'f'},
		'4': {'g', 'h', 'i'},
		'5': {'j', 'k', 'l'},
		'6': {'m', 'n', 'o'},
		'7': {'p', 'q', 'r', 's'},
		'8': {'t', 'u', 'v'},
		'9': {'w', 'x', 'y', 'z'},
	}

	var backtrack func(index int, current []byte)
	backtrack = func(index int, current []byte) {
		if index == len(digits) {
			paths = append(paths, string(current))
			return
		}

		letters := m[digits[index]]
		for i := 0; i < len(letters); i++ {
			backtrack(index+1, append(current, letters[i]))
		}
	}
	backtrack(0, []byte{})
	return paths
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L17电话号码的字母组合.java
package top.dyw.leetcode.Hot100;

public class L17电话号码的字母组合 {
    List<String> result = new ArrayList();
    String[] Mapping = new String[]{
            "",
            "",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
    };
    public List<String> letterCombinations(String digits) {

        dfs(digits, 0, new StringBuilder());
        return result;
    }

    public void dfs(String digits, int index, StringBuilder path) {
        if (path.length() == digits.length()) {
            result.add(path.toString());
            return;
        }

        String letters = Mapping[digits.charAt(index)-'0'];
        for (int i=0; i<letters.length(); i++) {
            path.append(letters.charAt(i));
            dfs(digits, index+1, path);
            path.deleteCharAt(path.length()-1);
        }
    }
}
